### ISSUE TITLE
TypeError: 'unicode' does not have the buffer interface

### ISSUE THREAD

**Issue by sh4nks** [cryptojuice/flask-scrypt#6]
I'm using Python 2.7.8 with all packages up-to-date. My mate is using Python 2.7.6 and didn't get this error. I think this issue is similar to the one that werkzeug has: https://github.com/mitsuhiko/werkzeug/issues/537.

    127.0.0.1 - - [12/Jul/2014 11:30:35] "POST /auth/login HTTP/1.1" 500 -
    Traceback (most recent call last):
      File "/home/peter/.virtualenvs/clean/lib/python2.7/site-packages/flask/app.py", line 1836, in __call__
        return self.wsgi_app(environ, start_response)
      File "/home/peter/.virtualenvs/clean/lib/python2.7/site-packages/flask/app.py", line 1820, in wsgi_app
        response = self.make_response(self.handle_exception(e))
      File "/home/peter/.virtualenvs/clean/lib/python2.7/site-packages/flask/app.py", line 1403, in handle_exception
        reraise(exc_type, exc_value, tb)
      File "/home/peter/.virtualenvs/clean/lib/python2.7/site-packages/flask/app.py", line 1817, in wsgi_app
        response = self.full_dispatch_request()
      File "/home/peter/.virtualenvs/clean/lib/python2.7/site-packages/flask/app.py", line 1477, in full_dispatch_request
        rv = self.handle_user_exception(e)
      File "/home/peter/.virtualenvs/clean/lib/python2.7/site-packages/flask/app.py", line 1381, in handle_user_exception
        reraise(exc_type, exc_value, tb)
      File "/home/peter/.virtualenvs/clean/lib/python2.7/site-packages/flask/app.py", line 1475, in full_dispatch_request
        rv = self.dispatch_request()
      File "/home/peter/.virtualenvs/clean/lib/python2.7/site-packages/flask_debugtoolbar/__init__.py", line 125, in dispatch_request
        return view_func(**req.view_args)
      File "/home/peter/Development/flaskbb/flaskbb/auth/views.py", line 37, in login
        form.password.data)
      File "/home/peter/Development/flaskbb/flaskbb/user/models.py", line 217, in authenticate
        authenticated = user.check_password(password)
      File "/home/peter/Development/flaskbb/flaskbb/user/models.py", line 201, in check_password
        return check_password_hash(password.encode("utf-8"), self.password, self._get_salt())
      File "/home/peter/.virtualenvs/clean/lib/python2.7/site-packages/flask_scrypt.py", line 125, in check_password_hash
        return safe_str_cmp(password_hash, candidate_hash)
      File "/home/peter/.virtualenvs/clean/lib/python2.7/site-packages/flask_scrypt.py", line 137, in safe_str_cmp
        return _builtin_safe_str_cmp(a, b)
    TypeError: 'unicode' does not have the buffer interface

Quote from a comment on the werkzeug issue:
>**shuhaowu** commented on 4 Jun
>Found the issue: On https://github.com/mitsuhiko/werkzeug/blob/master/werkzeug/security.py#L133 it uses the builtin hmac.compare_digest to compare. This is a new method in 2.7.7. The documentation shows that it needs to compare two items that have the same type.

>https://docs.python.org/2/library/hmac.html#hmac.compare_digest

**Comment by hivre** [comment]
I've never made a pull-request, so don't how to do that, but do know how to solve this. There are two ways:

either wise add:
``` python
    if isinstance(a, text_type):
        a = a.encode('utf-8')
    if isinstance(b, text_type):
        b = b.encode('utf-8')
```

On line 137, but that means importing `text_type` from Werkzeug. (I am not sure how they define it there)

or 

remove the `safe_str_cmp()` function, and use Werkzeug's.  (by importing `from werkzeug.security import safe_str_cmp`)

The latter seems prudent, as the Scrypt function is the same as Werkzeugs. However, not sure what that does to speed/performance.

**Comment by cryptojuice** [comment]
Thanks @hivre I will test your fixes and update the package.

**Comment by Iker-Jimenez** [comment]
Looks like this fix was never released. 
Would it be possible to get a new release out with this fix? Latest I can find in PIP is 0.1.3.6, which still uses its own implementation of safe_str_cmp instead of the one in werkzeug.security

**Comment by Sandvich** [comment]
Please get this update onto pypi asap, I'm having exactly this issue and have had to manually change the file which is not what I want to be doing long term. Thank you! :)
